core/src/sigverify.rs

```diff
@@ -81,23 +85,55 @@ fn batch_size(batches: &[Packets]) -> usize {
     batches.iter().map(|p| p.packets.len()).sum()
 }
 
-pub fn get_packet_offsets(packet: &Packet, current_offset: u32) -> (u32, u32, u32, u32) {
-    let (sig_len, sig_size) = decode_len(&packet.data);
-    let msg_start_offset = sig_size + sig_len * size_of::<Signature>();
+// internal function to be unit-tested; should be used only by get_packet_offsets
+fn do_get_packet_offsets(packet: &Packet, current_offset: u32) -> (bool, u32, u32, u32, u32) {
+    // This directly reads the length of Transaction.signatures (serialized with short_vec)
+    let (sig_len_untrusted, sig_size) = decode_len(&packet.data);
+
+    // This directly reads MessageHeader.num_required_signatures (serialized with u8)
+    let msg_start_offset = sig_size + sig_len_untrusted * size_of::<Signature>();
+    // Using msg_start_offset which is based on sig_len_untrusted introduces uncertainity.
+    // Ultimately, the actual sigverify will determine the uncertainity.
+    let sig_len_maybe_trusted = packet.data[msg_start_offset] as usize;
 
-    let (_pubkey_len, pubkey_size) = decode_len(&packet.data[msg_start_offset..]);
```

Comment:
Sadly, this was a blatant wrongdoing of reading a packet...

In short, this old code incorrectly reads bytes serialized from `MessageHeader` (three of `u8`s) as the length (`ShortU16`) of `short_vec`. (Moderate pun intended... :).

So, if `MessageHeader.required_num_sigs` are above `0x7f` (according to the comment of `Short16`), `pubkey_size` will be `2` and subsequent offset calculation gets bogus, resulting in **false negative results of sigverify**. This have been working because most of time `MessageHeader.required_num_sigs` is very small compared to the `0x7f`.

To make sure, I traced this back to the origin of introduction of this bug. Let me know if it's needed!

(By the way, this is the first bug I've ever found in the production code of solana, I think I've finally managed to make tangible contribution... yay!)

Reply:
Also, accompanying unit test is `test_large_sigs`

Reply:
This is a nice catch. That was my oversight; thanks!

Reply:
FYI, this is split into #6388.